src/libraries/System.Private.CoreLib/src/System/DateTimeOffset.Android.cs

```diff
@@ -0,0 +1,69 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Threading;
+
+namespace System
+{
+    public readonly partial struct DateTimeOffset
+    {
+        // 0 == in process of being loaded, 1 == loaded
+        private static int s_androidTZDataLoaded = -1;
+
+        // Now on Android does the following
+        // 1) quickly returning a fast path result when first called if the right AppContext data element is set
+        // 2) starting a background thread to load TimeZoneInfo local cache
+        //
+        // On Android, loading AndroidTZData is expensive for startup performance.
+        // The fast result relies on `System.TimeZoneInfo.LocalDateTimeOffset` being set
+        // in the App Context's properties as the appropriate local date time offset from UTC.
+        // monovm_initialize(_preparsed) can be leveraged to do so.
+        // However, to handle timezone changes during the app lifetime, AndroidTZData needs to be loaded.
+        // So, on first call, we return the fast path and start a background thread to load
+        // the TimeZoneInfo Local cache implementation which loads AndroidTZData.
+        public static DateTimeOffset Now
+        {
+            get
+            {
+                DateTime utcDateTime = DateTime.UtcNow;
+
+                if (s_androidTZDataLoaded == 1) // The background thread finished, the cache is loaded.
+                {
+                    return ToLocalTime(utcDateTime, true);
+                }
+
+                // The cache isn't loaded yet.
+                if (Interlocked.CompareExchange(ref s_androidTZDataLoaded, 0, -1) == -1)
+                {
+                    new Thread(() =>
+                    {
+                        try
+                        {
+                            // Delay the background thread to avoid impacting startup, if it still coincides after 1s, startup is already perceived as slow
+                            Thread.Sleep(1000);
+
+                            _ = TimeZoneInfo.Local; // Load AndroidTZData
+                        }
+                        finally
+                        {
+                            Volatile.Write(ref s_androidTZDataLoaded, 1);
+                        }
+                    }) { IsBackground = true }.Start();
+                }
+
+                object? localDateTimeOffset = AppContext.GetData("System.TimeZoneInfo.LocalDateTimeOffset");
+                if (localDateTimeOffset == null) // If no offset property provided through monovm app context, default
+                    return ToLocalTime(utcDateTime, true);
```

Comment:
I hadn't looked at this logic before.  If localDateTimeOffset is null, does it still make sense to be creating a thread, since ToLocalTime presumably is about to aggressively load the state?  I'm wondering if these three lines should be moved up to before the CompareExchange.

Reply:
e.g.
```C#
            get
            {
                DateTime utcDateTime = DateTime.UtcNow;
                object? localDateTimeOffset;

                if (s_androidTZDataLoaded == 1 || // The background thread finished, the cache is loaded.
                    (localDateTimeOffset = AppContext.GetData("System.TimeZoneInfo.LocalDateTimeOffset")) is null)
                {
                    return ToLocalTime(utcDateTime, true);
                }

                ...
```

Reply:
Other than that, the threading logic looks good now.

Reply:
Yes, I think it still makes sense.  We still need to prevent tzdata from being loaded during startup.

Reply:
> We still need to prevent tzdata from being loaded during startup.

Won't calling `ToLocalTime(utcDateTime, true)` like this load tzdata?

Reply:
Yes, you're right. I updated the PR.